File: CommandNavigation/CommandManualNavigationx2.cs
namespace CommandNavigation.Command1Navigation2 {
  using System;
  using System.Collections.Generic;
  using System.Text;

  public class CommandManualNavigation<T> : Stack<T> where T : class, ICommandManualCtrl {
    public CommandManualNavigation() : base() { }
    public CommandManualNavigation(int Capacity) : base(Capacity) { }

    public event OnCommandManualPoppedHandle<T> OnPopped;
    public event OnCommandManualPushedHandle<T> OnPushed;

    public new void Clear() {
      while (Count > 0) {
        this.Pop();
      }
    }
    public void Discard() { base.Clear(); }

    public new void Push(T Item) {
      base.Push(Item);
      Item.OnPush();
      OnPushed?.Invoke(this, Item);
    }
    public new T Pop() {
      var Item = base.Pop();
      Item.OnPop();
      OnPopped?.Invoke(this, Item);
      return Item;
    }

  }
  public delegate void OnCommandManualPushedHandle<T>(CommandManualNavigation<T> CurrentStack, T PushedItem) where T : class, ICommandManualCtrl;
  public delegate void OnCommandManualPoppedHandle<T>(CommandManualNavigation<T> CurrentStack, T PoppedItem) where T : class, ICommandManualCtrl;
}

File: CommandNavigation.TS/src/CommandManualNavigationx2.ts
import { ICommandManualCtrl } from "./Interfaces";
import { Event2 } from "./../node_modules/katkits.ts/lib/Event"

export default class CommandManualNavigation extends Array<ICommandManualCtrl>{
  public readonly OnPopped: Event2<CommandManualNavigation, ICommandManualCtrl>;
  public readonly OnPushed: Event2<CommandManualNavigation, ICommandManualCtrl>;

  constructor() {
    super();
    this.OnPopped = new Event2();
    this.OnPushed = new Event2();
  }

  public Peek(): ICommandManualCtrl { return this[this.length - 1]; }

  public Push(Item: ICommandManualCtrl): void {
    super.push(Item);
    Item.OnPush();
    this.OnPushed.Invoke(this, Item);
  }

  public Pop(): ICommandManualCtrl {
    if (this.length > 0) {
      const Popped = super.pop();
      Popped.OnPop();
      this.OnPopped.Invoke(this, Popped);
      return Popped;
    }
    throw Error("Empty Stack");
  }

  public Clear() {
    while (this.length > 0) {
      this.Pop();
    }
  }
  public Discard() { super.splice(0, this.length); }


}
